File: serverless-ecommerce/src/products/get-product.js
console.log("Loading get product function");

const AWS = require("aws-sdk");
AWS.config.update({ region: "eu-west-2" });

const dynamodb = new AWS.DynamoDB.DocumentClient();

exports.handler = async (event, context) => {
  console.log("Received event:", JSON.stringify(event, null, 2));

  try {
    console.log("PRODUCTS_TABLE:", process.env.PRODUCTS_TABLE);

    if (!process.env.PRODUCTS_TABLE) {
      throw new Error("PRODUCTS_TABLE environment variable is not set");
    }

    const { productId } = event.pathParameters || {};
    console.log("Product ID:", productId);

    if (!productId) {
      return {
        statusCode: 400,
        headers: { "Access-Control-Allow-Origin": "*" },
        body: JSON.stringify({
          success: false,
          error: "Product ID is required in the URL path",
        }),
      };
    }

    const params = {
      TableName: process.env.PRODUCTS_TABLE,
      Key: { productId },
    };

    console.log("Get item params:", JSON.stringify(params));

    const result = await dynamodb.get(params).promise();
    console.log("Get item result:", JSON.stringify(result));

    if (!result.Item) {
      return {
        statusCode: 404,
        headers: { "Access-Control-Allow-Origin": "*" },
        body: JSON.stringify({
          success: false,
          error: `Product with ID ${productId} not found`,
        }),
      };
    }

    return {
      statusCode: 200,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        success: true,
        product: result.Item,
      }),
    };
  } catch (error) {
    console.error("FUNCTION ERROR:", error);
    console.error("Error name:", error.name);
    console.error("Error message:", error.message);
    console.error("Error code:", error.code);

    return {
      statusCode: 500,
      headers: { "Access-Control-Allow-Origin": "*" },
      body: JSON.stringify({
        success: false,
        error: "Internal server error",
        message: error.message,
        code: error.code,
      }),
    };
  }
};

File: terraform/lambdas/products/get_product.py

import json
import boto3
import os

dynamodb = boto3.resource('dynamodb')
products_table = dynamodb.Table(os.environ['PRODUCTS_TABLE'])

def lambda_handler(event, context):
    try:
        product_id = event['pathParameters']['product_id']
        
        response = products_table.get_item(Key={'productId': product_id})
        
        if 'Item' not in response:
            return {
                'statusCode': 404,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({'error': 'Product not found'})
            }
        
        product = response['Item']
        
        # Convert Decimal to float
        if 'price' in product:
            product['price'] = float(product['price'])
        if 'stock_quantity' in product:
            product['stock_quantity'] = int(product['stock_quantity'])
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps(product)
        }
        
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({'error': str(e)})
        }
